<a href="https://colab.research.google.com/github/byui-cse/cse380-notebooks/blob/master/10_3_About_Patterns_and_Probabilities.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# About Patterns and Probabilities
## Class Directed Learning
### Due: Tuesday, 9 March 2021, 11:59 pm

## TODO Explore and Wonder

Regarding spanning trees of ladder graphs:

What is the probability ($\lim_{n \rightarrow \infty} P(n)$) that a randomly-selected spanning tree of an $n$-rung ladder graph contains the bottom rung?

| n | P(n) |
|---|------|
| 1 |      |
| 2 |      |
| 3 |      |
| 4 |      |
| 5 |      |


### Recurrence Relations

Let $f(n) =$ NSTIBR$(n)$:

$f(n) = 4f(n-1) - f(n-2)$ for $n > 1$;

$f(0) = 1$,

$f(1) = 1$.

Let $g(n) =$ NST$(n)$:

$g(n) = 4g(n-1) - g(n-2)$ for $n > 1$;

$g(0) = 0$,

$g(1) = 1$.

#### TODO Check Recurrences

Check the calculations in this table (maybe write recursive functions) to verify they are accurate.

Do they match what you found yesterday in your DPC?

| n | f(n) | f(n-1) | 4f(n-1)  | f(n - 2) | diff |
|---|-----:|-------:|---------:|---------:|-----:|
| 0 |    1 |    N/A |      N/A |      N/A |  N/A |
| 1 |    1 |      1 |        4 |      N/A |  N/A |
| 2 |    3 |      1 |        4 |        1 |    3 |
| 3 |   11 |      3 |       12 |        1 |   11 |
| 4 |   41 |     11 |       44 |        3 |   41 |
| 5 |  153 |     41 |      164 |       11 |  153 |

| n | g(n) | g(n-1) | 4g(n-1)  | g(n - 2) | diff |
|---|-----:|-------:|---------:|---------:|-----:|
| 0 |    0 |    N/A |      N/A |      N/A |  N/A |
| 1 |    1 |      0 |        0 |      N/A |  N/A |
| 2 |    4 |      1 |        4 |        0 |    4 |
| 3 |   15 |      4 |       16 |        1 |   15 |
| 4 |   56 |     15 |       60 |        4 |   56 |
| 5 |  209 |     56 |      224 |       15 |  209 |

### TODO Find closed-form formulas

Can you find closed-form formulas for $f(n)$ and $g(n)$?

A closed-form formula expressing these functions in terms of operations on $n$, **without** referring to previous calculated values of the functions.

In [2]:
import math

def NSTCount(n):
    return (1/ (2 * math.sqrt(3)) * (2 + math.sqrt(3)) ** n - 1/ (2 * math.sqrt(3)) * (2 - math.sqrt(3)) ** n)

def closed_NSTIBR(n):
  return int(1/ (2 * math.sqrt(3)) * (2 + math.sqrt(3)) ** n - 1/ (2 * math.sqrt(3)) * (2 - math.sqrt(3)) ** n) - int(1/ (2 * math.sqrt(3)) * (2 + math.sqrt(3)) ** (n-1) - 1/ (2 * math.sqrt(3)) * (2 - math.sqrt(3)) ** (n-1))


# Driver code
for n in range(1, 50):
    NST = math.floor((NSTCount(n)))
    NSTIBR = closed_NSTIBR(n)
    print(f"NST: {n}" )
    print(f"NST: {math.floor(NSTCount(n))}" )
    print(f"NSTIBR: {math.floor(closed_NSTIBR(n))}\n" )


NST: 1
NST: 1
NSTIBR: 1

NST: 2
NST: 4
NSTIBR: 3

NST: 3
NST: 15
NSTIBR: 11

NST: 4
NST: 56
NSTIBR: 41

NST: 5
NST: 209
NSTIBR: 153

NST: 6
NST: 780
NSTIBR: 571

NST: 7
NST: 2911
NSTIBR: 2131

NST: 8
NST: 10863
NSTIBR: 7952

NST: 9
NST: 40545
NSTIBR: 29682

NST: 10
NST: 151315
NSTIBR: 110770

NST: 11
NST: 564718
NSTIBR: 413403

NST: 12
NST: 2107559
NSTIBR: 1542841

NST: 13
NST: 7865520
NSTIBR: 5757961

NST: 14
NST: 29354523
NSTIBR: 21489003

NST: 15
NST: 109552574
NSTIBR: 80198051

NST: 16
NST: 408855775
NSTIBR: 299303201

NST: 17
NST: 1525870528
NSTIBR: 1117014753

NST: 18
NST: 5694626339
NSTIBR: 4168755811

NST: 19
NST: 21252634830
NSTIBR: 15558008491

NST: 20
NST: 79315912983
NSTIBR: 58063278153

NST: 21
NST: 296011017104
NSTIBR: 216695104121

NST: 22
NST: 1104728155435
NSTIBR: 808717138331

NST: 23
NST: 4122901604638
NSTIBR: 3018173449203

NST: 24
NST: 15386878263119
NSTIBR: 11263976658481

NST: 25
NST: 57424611447840
NSTIBR: 42037733184721

NST: 26
NST: 214311567528243
NSTIBR: 156

#### Hint:

http://www.ist.tugraz.at/aichholzer/teaching/eca/spanning_trees_in_ladders.pdf

#### Possibly Illuminating Calculations

Remember continued fractions?

In [1]:
from fractions import Fraction as frac

def contfrac2frac(seq):
    """Convert the simple continued fraction in `seq`
       into a fraction with numerator num and denominator den.
    """
    num, den = 1, 0
    for u in reversed(seq):
        num, den = den + num * u, num
    return frac(num, den)

def frac2contfrac(f):
    """Build the simple continued fraction expansion of fraction f.
    """
    seq = []
    frac2contfrac_rec(f, seq)
    return seq

def frac2contfrac_rec(f, seq):
    n = f.numerator
    d = f.denominator
    if d != 0:
        seq.append(n // d)
        if n % d != 0:
            frac2contfrac_rec(frac(d, n % d), seq)

def eval_frac(f):
    """Evaluate the fraction f as a float.
    """
    return f.numerator / f.denominator

In [ ]:
from math import sqrt

value = sqrt(3) - 1
value_as_cf = frac2contfrac(frac.from_float(value))[:21]
cf_to_value = contfrac2frac(value_as_cf)
print(value, value_as_cf, cf_to_value)
print(eval_frac(cf_to_value))

In [ ]:
from pprint import pprint

pprint(list(map(lambda n: contfrac2frac(value_as_cf[:n]), range(20, 5, -1))))

In [4]:
import math

def NSTCount(n):
    return (1/ (2 * math.sqrt(3)) * (2 + math.sqrt(3)) ** n - 1/ (2 * math.sqrt(3)) * (2 - math.sqrt(3)) ** n)

def closed_NSTIBR(n):
  return int(1/ (2 * math.sqrt(3)) * (2 + math.sqrt(3)) ** n - 1/ (2 * math.sqrt(3)) * (2 - math.sqrt(3)) ** n) - int(1/ (2 * math.sqrt(3)) * (2 + math.sqrt(3)) ** (n-1) - 1/ (2 * math.sqrt(3)) * (2 - math.sqrt(3)) ** (n-1))


# Driver code
for n in range(1, 100):
    NST = math.floor((NSTCount(n)))
    NSTIBR = closed_NSTIBR(n)
    print(f"NSTIBR / NST: {NSTIBR / NST}\n" )

NSTIBR / NST: 1.0

NSTIBR / NST: 0.75

NSTIBR / NST: 0.7333333333333333

NSTIBR / NST: 0.7321428571428571

NSTIBR / NST: 0.7320574162679426

NSTIBR / NST: 0.732051282051282

NSTIBR / NST: 0.7320508416351769

NSTIBR / NST: 0.7320261437908496

NSTIBR / NST: 0.7320754716981132

NSTIBR / NST: 0.7320490367775832

NSTIBR / NST: 0.7320521038819375

NSTIBR / NST: 0.7320511549142871

NSTIBR / NST: 0.7320509006397543

NSTIBR / NST: 0.7320508325071404

NSTIBR / NST: 0.7320508142510646

NSTIBR / NST: 0.732050809359364

NSTIBR / NST: 0.7320508080486368

NSTIBR / NST: 0.7320508076974285

NSTIBR / NST: 0.7320508076033224

NSTIBR / NST: 0.7320508075781068

NSTIBR / NST: 0.7320508075713503

NSTIBR / NST: 0.73205080756954

NSTIBR / NST: 0.7320508075690548

NSTIBR / NST: 0.7320508075689248

NSTIBR / NST: 0.7320508075688901

NSTIBR / NST: 0.7320508075688807

NSTIBR / NST: 0.7320508075688782

NSTIBR / NST: 0.7320508075688774

NSTIBR / NST: 0.7320508075688773

NSTIBR / NST: 0.7320508075688773

NSTIBR / NST:

The percentage of the growth of g(n) is always around 73% when it is iterating when n > 3.
Therefore when we use (g(n)- g(n-1))/ g(n), when n -> infinite. It will be almost equal to 73%.
![title](./spanningTreeDPC10.png)
